src/style/style.js

```diff
@@ -937,44 +942,81 @@ class Style extends Evented {
-            }
-        }
-        return false;
-    }
-
     _generateCollisionBoxes() {
         for (const id in this.sourceCaches) {
             this._reloadSource(id);
         }
     }
 
     _updatePlacement(transform: Transform, showCollisionBoxes: boolean, fadeDuration: number) {
-        const forceFullPlacement = this.getNeedsFullPlacement();
+        let symbolBucketsChanged = false;
+        let placementChanged = false;
+
+        const layerTiles = {};
+
+        for (const layerID of this._order) {
+            const styleLayer = this._layers[layerID];
+            if (styleLayer.type !== 'symbol') continue;
+
+            if (!layerTiles[styleLayer.source]) {
+                const sourceCache = this.sourceCaches[styleLayer.source];
+                layerTiles[styleLayer.source] = sourceCache.getRenderableIds()
+                    .map((id) => sourceCache.getTileByID(id))
+                    .sort((a, b) => (b.tileID.overscaledZ - a.tileID.overscaledZ) || (a.tileID.isLessThan(b.tileID) ? -1 : 1));
+            }
 
-        if (forceFullPlacement || !this.placement || this.placement.isDone()) {
-            this.placement = new Placement(transform, this._order, forceFullPlacement, showCollisionBoxes, fadeDuration, this.placement);
+            const layerBucketsChanged = this.crossTileSymbolIndex.addLayer(styleLayer, layerTiles[styleLayer.source]);
+            symbolBucketsChanged = symbolBucketsChanged || layerBucketsChanged;
+        }
+
+        // Anything that changes our "in progress" layer and tile indices requires us
+        // to start over. When we start over, we do a full placement instead of incremental
+        // to prevent starvation.
+        // We need to restart placement to keep layer indices in sync.
+        const forceFullPlacement = this._layerOrderChanged;
+
+        if (forceFullPlacement || !this.pauseablePlacement || (this.pauseablePlacement.isDone() && !this.placement.stillRecent(browser.now()))) {
+            this.pauseablePlacement = new PauseablePlacement(transform, this._order, forceFullPlacement, showCollisionBoxes, fadeDuration, layerTiles);
             this._layerOrderChanged = false;
         }
 
-        this.placement.continuePlacement(this._order, this._layers, this.sourceCaches);
+        if (!this.pauseablePlacement.isDone()) {
+            this.pauseablePlacement.continuePlacement(this._order, this._layers, layerTiles);
+
+            if (this.pauseablePlacement.isDone()) {
+                const placement = this.pauseablePlacement.placement;
+                placementChanged = placement.commit(this.placement, browser.now());
+                if (!this.placement || placementChanged || symbolBucketsChanged) {
+                    this.placement = placement;
+                    this.collisionIndex = this.placement.collisionIndex;
+                    this.placement.setRecent(browser.now());
+                } else {
+                    this.placement.setRecent(browser.now());
```

Comment:
This was tricky for me to read -- it looks like `setRecent` is just being duplicated, it's easy to miss that `setRecent` is going to set `stale` to false, and confusing that we want to undo it right away (because we want to update `_recentUntil` for throttling, right?).

Could we just save something like `inProgressPlacementStale` at the top and set it at the bottom? It might be clearer.

Or would it make sense to simplify this just by always committing the pausable placement? Do we expect there's some performance win from keeping the original placement object unmodified in the case nothing changes?